File: python/longest_substring_without_repeating_characters.py
class Solution:
    # @param a, a string
    # @param b, a string
    # @return a string
    def lengthOfLongestSubstring(self, s):
        """
        Declare a list to hold characters in current examined substring.
        If next character in s occured in current substring, pop up characters 
        from the beginning of current substring until we poped up the duplicated 
        character.
        """
        hash_list = [False] * 1000 #initilizes a list. The index of the list is the ascii of the character, the value indicates whether this character contains in current examined substring.
        max_len = -1
        r = [] # current examined substring.
        for c in s:
            ascii_of_c = ord(c)
            if hash_list[ascii_of_c]:
                if len(r) > max_len:
                    max_len = len(r)

                while r[0] != c:
                    hash_list[ord(r[0])] = False
                    r.pop(0)
                    
                r.pop(0)
            hash_list[ascii_of_c] = True
            r.append(c)

        if len(r) > max_len:
            max_len = len(r)
        return max_len

if __name__ == '__main__':
    l = ['abcabcbb', 'bbbbbbbbb', '', 'wlrbbmqbhcdarzowkkyhiddqscdxrjmowfrxsjybldbefsarcbynecdyggxxpklorellnmpapqfwkhopkmco']
    for ele in l:
        print ele, Solution().lengthOfLongestSubstring(ele)

File: c++/longest_substring_without_repeating_characters.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int max_len = 0, start = 0;
        int l = s.size();
        vector<int> idx(256, -1);
        for (int i = 0; i != l; i++) {
          //int c = s[i] - 'a';
          const unsigned char c=(unsigned char)s[i];
          int j = idx[c];
          idx[c] = i;
          //cout << "i:" << i << " s[i]:" << s[i] << " c:" << c << " j:" << j;
          if (start <= j) {
            start = j+1;
          }
          max_len = max(max_len, i-start+1);
          //cout << " start:" << start << " max_len:" << max_len << endl;
        }
        return max_len;
    }
};


int main()
{
    Solution *s = new Solution();
    int r;
    string t1("abcabcbb");
    r = s->lengthOfLongestSubstring(t1);
    cout << r << endl;
    string t2 = "bbbbbbbbb";
    r = s->lengthOfLongestSubstring(t2);
    cout << r << endl;
    string t3 = "";
    r = s->lengthOfLongestSubstring(t3);
    cout << r << endl;
    string t4 = "wlrbbmqbhcdarzowkkyhiddqscdxrjmowfrxsjybldbefsarcbynecdyggxxpklorellnmpapqfwkhopkmco";
    r = s->lengthOfLongestSubstring(t4);
    cout << r << endl;
    return 0;
}
